### ISSUE TITLE
NaN loss when training on single node multi-GPUs in AMP mode

### ISSUE THREAD

**Issue by ymwangg** [pytorch/xla#3200]
## 🐛 Bug

We saw `NaN` loss when training with single node multi-GPUs in AMP mode.

## To Reproduce

On example is the `test_train_mp_imagenet.py` training script with real dataset.

1. Pull the nightly image and launch it with multiple GPUs.
```
docker pull gcr.io/tpu-pytorch/xla:nightly_3.7_cuda_11.2
docker run --rm -it --gpus device=all -v /home/ubuntu/dataset/ILSVRC/Data/CLS-LOC:/dataset --entrypoint /bin/bash --ipc=host gcr.io/tpu-pytorch/xla:nightly_3.7_cuda_11.2
```
2. Run the imagenet test case
```
cd /pytorch/xla/test
GPU_NUM_DEVICES=4 python test_train_mp_imagenet.py --datadir /dataset/ --amp --log_steps 1
```
You will see `NaN` loss within a few hundred steps.
Below is an example output. Please note I manually logged if the optimizer step is taken in the GradScaler.
```
| Training Device=xla:0/2 Epoch=1 Step=55 Loss=7.02240 Rate=664.64 GlobalRate=78.53 Time=21:10:28
| Training Device=xla:0/3 Epoch=1 Step=55 Loss=7.00317 Rate=648.86 GlobalRate=78.53 Time=21:10:28
| Training Device=xla:0/1 Epoch=1 Step=55 Loss=7.00562 Rate=660.16 GlobalRate=78.52 Time=21:10:28
| Training Device=xla:0/0 Epoch=1 Step=55 Loss=7.08817 Rate=545.63 GlobalRate=78.35 Time=21:10:28
step@1
step@3
step@2
step@0
| Training Device=xla:0/2 Epoch=1 Step=56 Loss=7.03790 Rate=325.13 GlobalRate=78.81 Time=21:10:29
| Training Device=xla:0/3 Epoch=1 Step=56 Loss=7.02115 Rate=318.72 GlobalRate=78.81 Time=21:10:29
| Training Device=xla:0/1 Epoch=1 Step=56 Loss=7.09274 Rate=323.15 GlobalRate=78.80 Time=21:10:29
| Training Device=xla:0/0 Epoch=1 Step=56 Loss=6.97351 Rate=277.02 GlobalRate=78.63 Time=21:10:29
skipped@3
skipped@0
skipped@2
skipped@1
| Training Device=xla:0/2 Epoch=1 Step=57 Loss=7.01929 Rate=469.35 GlobalRate=80.00 Time=21:10:29
| Training Device=xla:0/3 Epoch=1 Step=57 Loss=6.91931 Rate=469.89 GlobalRate=80.00 Time=21:10:29
| Training Device=xla:0/1 Epoch=1 Step=57 Loss=6.95343 Rate=474.06 GlobalRate=79.99 Time=21:10:29
| Training Device=xla:0/0 Epoch=1 Step=57 Loss=2306.75781 Rate=655.02 GlobalRate=79.89 Time=21:10:29
skipped@0
skipped@1
skipped@2
skipped@3
| Training Device=xla:0/0 Epoch=1 Step=58 Loss=1528.33398 Rate=780.74 GlobalRate=81.13 Time=21:10:30
| Training Device=xla:0/2 Epoch=1 Step=58 Loss=7.00317 Rate=702.77 GlobalRate=81.25 Time=21:10:30
| Training Device=xla:0/1 Epoch=1 Step=58 Loss=7.04694 Rate=704.69 GlobalRate=81.24 Time=21:10:30
| Training Device=xla:0/3 Epoch=1 Step=58 Loss=7.07901 Rate=478.69 GlobalRate=81.15 Time=21:10:30
skipped@1
skipped@2
skipped@0
skipped@3
| Training Device=xla:0/2 Epoch=1 Step=59 Loss=6.96243 Rate=575.93 GlobalRate=82.39 Time=21:10:30
| Training Device=xla:0/1 Epoch=1 Step=59 Loss=7.06561 Rate=576.67 GlobalRate=82.38 Time=21:10:30
| Training Device=xla:0/3 Epoch=1 Step=59 Loss=7.03732 Rate=718.70 GlobalRate=82.39 Time=21:10:30
| Training Device=xla:0/0 Epoch=1 Step=59 Loss=nan Rate=605.51 GlobalRate=82.28 Time=21:10:30
```
 You can see at step 56, `NaN` suddenly occurrs in the gradients and `loss@device0 = 2306.75781` seems odd that suggests the model may be corrupted.

Notably, we found training with the default `torch.utils.data.DataLoader` seems to solve this issue with `data = data.to(device); target = target.to(device)` added in the training loop. Therefore this issue may be related to `pl.MpDeviceLoader`.

## Environment
Tested with 4 Nvidia V100 GPUs.

## Additional context
Here are some other useful observations:
1. This issue only occurs when training with real data in AMP mode. Training with fake data or FP32 seems fine.
~~2. In another training script with the same `NaN` loss issue, the `found_inf` tensor in the GradScaler starts to diverge after a few steps and thus some GPUs accumulates `NaN` gradients due to the wrong `found_inf` tensor.~~
3. This issue also occurs with `--use_zero_grad` and `--use_syncfree_optim`. Though it seems `NaN` occurs much later when using these two options.

Based on these observations, I suspect this issue may be due to GPU memory corruption where the model weights / ~~found_inf tensor~~ were tampered for some reason. But if `pl.MpDeviceLoader` causes the memory corruption, I'm not sure why this only happens in AMP mode.
cc @JackCaoG @yaochengji @codeislife99

### Update 11-04:
The `found_inf` tensor difference in observation 2 is due to a bug in our codebase. Sorry for the confusion.

**Comment by shauheen** [comment]
Thank you for reporting this issue, we will work on reproducing it on our side.

**Comment by miladm** [comment]
Hi @ymwangg. A

 follow up to your comment "But if pl.MpDeviceLoader causes the memory corruption, I'm not sure why this only happens in AMP mode."

From the repro comments, it sseems you tried FP32 on fake data only. Can you clarify if you also tried FP32 on real data?

**Comment by ymwangg** [comment]
@miladm FP32 works fine with both fake data and real data.
I've done additional tests. Interestingly, the AMP NaN loss issue seems to be gone after reducing the batch size to 32.
- [ ] NaN occurred `GPU_NUM_DEVICES=4 python test_train_mp_imagenet.py --datadir /dataset/ --num_epochs 1 --amp`
- [ ] NaN occurred `GPU_NUM_DEVICES=4 python test_train_mp_imagenet.py --datadir /dataset/ --num_epochs 1 --amp --batch_size 64`
- [x] Works `GPU_NUM_DEVICES=4 python test_train_mp_imagenet.py --datadir /dataset/ --num_epochs 1 --amp --batch_size 32`
- [x] Works `GPU_NUM_DEVICES=4 python test_train_mp_imagenet.py --datadir /dataset/ --num_epochs 1`
- [x] Works `GPU_NUM_DEVICES=4 python test_train_mp_imagenet.py --datadir /dataset/ --num_epochs 1 --batch_size 64`

**Comment by yaochengji** [comment]
Hi @ymwangg ,

I used the torch/xla amp to train Resnet about one year ago, the final accuracy is very simlilar to original pytorch's, with batch_size=256, GPU_NUM=8.

**Comment by miladm** [comment]
Hi @ymwangg to repro using the exact same data, would you please share your dataset dowload instructions?

**Comment by wonjoolee95** [comment]
Thanks for the response, @ymwangg. 

I've tested running with fake data in AMP mode and did not observe any `NaN` on my end. Ran with `GPU_NUM_DEVICES=4 python pytorch/xla/test/test_train_mp_imagenet.py --fake_data --num_epochs 1 --amp`. I'll follow-up with running with real data in AMP mode.

**Comment by ymwangg** [comment]
@yaochengji Thanks for the info.

After some investigation, I found the `NaN` loss issue is related to #3175. I suspect this is due to the tensorflow update. I tested the following images and the older one works.
```
NaN loss: gcr.io/tpu-pytorch/xla:nightly_3.7_cuda_11.2_20211024
Works: gcr.io/tpu-pytorch/xla:nightly_3.7_cuda_10.2_20211020
```

Another interesting thing I noticed is the image `nightly_3.7_cuda_10.2_20211020` still has `NaN` loss issue but occurs much less frequently (once every few hundred steps) and seems to be not affecting the training.

@miladm @wonjoolee95 I followed this tutorial to download the dataset https://github.com/pytorch/examples/blob/master/imagenet/README.md.

```
# download from kaggle https://www.kaggle.com/c/imagenet-object-localization-challenge/data
kaggle competitions download -c imagenet-object-localization-challenge
# unzip and run the following script
cd /home/ubuntu/dataset/ILSVRC/Data/CLS-LOC/val
wget wget https://raw.githubusercontent.com/soumith/imagenetloader.torch/master/valprep.sh
bash valprep.sh
```

**Comment by wonjoolee95** [comment]
Thanks for the detailed info, @ymwangg! I'll download the dataset and test it as soon as the download finishes.

**Comment by wonjoolee95** [comment]
Testing with real data with `--amp`, can confirm that NaN was observed. 
```
GPU_NUM_DEVICES=4 python pytorch/xla/test/test_train_mp_imagenet.py --datadir /dataset/ILSVRC/Data/CLS-LOC/ --num_epochs 1 --amp
```

First `NaN` was observed in around `Step=200`.

**Comment by wonjoolee95** [comment]
It seems like it works fine with the latest `cuda_10.2` (i.e., `nightly_3.7_cuda_10.2_20211020`) but does **not work** with the latest `cuda_11.2` (i.e., `nightly_3.7_cuda_11.2_20211106`). 

Upon some more investigation, no `NaN` was observed with `nightly_3.7_cuda_11.2_20211009`, so it appears there seems to be an issue between `nightly_3.7_cuda_11.2_20211009` and `nightly_3.7_cuda_11.2_20211106`.

**Comment by JackCaoG** [comment]
Hmm. interesting. We are actually thinking about deprecating cuda 10.2 due to known compilation issue with XLA:GPU and cuda 10.2. 

My guess is that `nan` came from TF update(https://github.com/pytorch/xla/pull/3175) like @ymwangg suggested. In Tf update we also update XLA:GPU so there is a behavior change for some ops.

Something we can try
1. Update tf again. Sometime this kind of bug just fix by xla:GPU folks when they noticed it. The problem of this approach is that Circle CI will fail since it is using cuda 10.2. Last time I tried the latest TF with cuda 10.2, it does not compile.
2. Try to bisect the code and find the exact op and input that caused the `nan`. This will likely be hard given it happens in epoch 200.

I think we can start with manually verify if update tf will solve the issue. @wonjoolee95 We can sync offline about how to do that.

**Comment by wonjoolee95** [comment]
I've tried updating the TF and training the model on `nightly_3.7_cuda_11.2_20211111`.

I was still observing sporadic `NaN` losses starting at `Step=1420` (much later compared to `Step=200` before updating the TF). Also this training was able to complete, while the training before updating the TF wasn't able to complete due to repeating `NaN` losses. 

@JackCaoG, how do you think we should proceed?

**Comment by JackCaoG** [comment]
@ymwangg Is Nan losses at step 1420 something that is acceptable to you?

**Comment by ymwangg** [comment]
@wonjoolee95 @JackCaoG Thanks for the debugging effort. Sporadic NaN that not affects training results should be acceptable. I think it's better to keep an eye on it but it's no longer critical.
@wonjoolee95 I found your fix (updating TF) works and I'm able to reproduce the sporadic NaN behavior. Additionally, I found reverting the following TF commit + using CUDA 11.4 (torch_xla==4863a3, tensorflow==e324b1) also works (sporadic NaN still exists).
```
commit 3fc86b593f342e910f6ba90e51b9511b71d1d9c1
Author: Artem Belevich <tra@google.com>
Date:   Thu Aug 12 11:53:38 2021 -0700

    Stabilize buffer assignment order.
    
    PiperOrigin-RevId: 390424917
    Change-Id: I22bb4ca4324735f236c2114a0b884ff411f4e9f9
```
Though this change looks trivial, I have no idea why this solves the NaN issue. Maybe there are subtle bugs on TF side.
